#### Import and preporcess fasting time and weight data from personal observations

In [21]:
import pandas as pd
from sqlalchemy import create_engine
import numpy as np

# Load your CSV data into a Pandas DataFrame
df = pd.read_csv('/home/turo/Downloads/data.csv', names=['time_delta', 'date_added', 'weight'])

In [22]:
df.head(20)

time_delta           date_added  weight
0         NaN   7/30/2019 20:25:00    77.0
1         NaN  12/30/1899 15:25:00     NaN
2         NaN   7/30/2019 20:25:00     NaN
3    24:35:00                  NaN     NaN
4         NaN                  NaN     NaN
5         NaN    2/8/2019 21:25:00     NaN
6         NaN    3/8/2019 15:45:00     NaN
7         NaN    4/8/2019 22:20:00     NaN
8         NaN    5/8/2019 14:15:00     NaN
9    24:00:00                  NaN     NaN
10        NaN       8/7/2019 17:03     NaN
11   17:07:00             8/8/2019    77.4
12        NaN             8/9/2019     NaN
13        NaN      8/10/2019 11:00     NaN
14   13:00:00    8/11/2019 2:10 AM     NaN
15   21:50:00   8/12/2019 10:00 AM     NaN
16   14:00:00   8/13/2019 12:36 PM     NaN
17   11:24:00   8/14/2019 11:30 PM     NaN
18    0:30:00            8/15/2019     NaN
19   33:30:00            8/16/2019     NaN

WATCH row in DF "11	17:07:00	8/8/2019	77.4" After conversion loses Date. COrrect before export to Posgres

In [11]:
# Handle Datetime issues
df['date_added'] = pd.to_datetime(df['date_added'], errors='coerce')
df.head(5)

time_delta          date_added  weight
0        NaN 2019-07-30 20:25:00    77.0
1        NaN 1899-12-30 15:25:00     NaN
2        NaN 2019-07-30 20:25:00     NaN
3   24:35:00                 NaT     NaN
4        NaN                 NaT     NaN

In [12]:
# Fasting data was calculated not always - 24.00 means there was no calculation 
# handle these entries
count_24_hours = (df['time_delta'] == '24:00:00').sum()
print("Count of '24:00:00' entries in time_delta column:", count_24_hours)


Count of '24:00:00' entries in time_delta column: 338


In [13]:
df.shape, df['time_delta'].isna().sum()

((1518, 3), 109)

In [14]:
df['time_delta'] = df['time_delta'].replace('24:00:00', np.nan)


In [16]:
count_24_hours = (df['time_delta'] == '24:00:00').sum()
print("Count of '24:00:00' entries in time_delta column:", count_24_hours)

print ('*************************')
df.shape, df['time_delta'].isna().sum()

Count of '24:00:00' entries in time_delta column: 0
*************************


((1518, 3), 447)

In [17]:
# Given this situation, one approach could be to manually parse the time_delta column to convert it 
# into a timedelta format. Since your time_delta column contains values like "24:35:00", 
# which represent hours:minutes:seconds, you can split the string and convert it into hours, 
# minutes, and seconds, and then create a timedelta object from these components.

import pandas as pd
import numpy as np

# Assuming df is your DataFrame
for idx, value in enumerate(df['time_delta']):
    if isinstance(value, str):
        parts = value.split(':')
        if len(parts) == 3:
            hours = int(parts[0])
            minutes = int(parts[1])
            seconds = int(parts[2])
            total_seconds = hours * 3600 + minutes * 60 + seconds
            df.at[idx, 'time_delta'] = pd.Timedelta(seconds=total_seconds)
        else:
            df.at[idx, 'time_delta'] = np.nan
    else:
        df.at[idx, 'time_delta'] = np.nan

In [20]:
df.head(20)

time_delta          date_added  weight
0               NaN 2019-07-30 20:25:00    77.0
1               NaN 1899-12-30 15:25:00     NaN
2               NaN 2019-07-30 20:25:00     NaN
3   1 days 00:35:00                 NaT     NaN
4               NaN                 NaT     NaN
5               NaN 2019-02-08 21:25:00     NaN
6               NaN 2019-03-08 15:45:00     NaN
7               NaN 2019-04-08 22:20:00     NaN
8               NaN 2019-05-08 14:15:00     NaN
9               NaN                 NaT     NaN
10              NaN                 NaT     NaN
11  0 days 17:07:00                 NaT    77.4
12              NaN                 NaT     NaN
13              NaN                 NaT     NaN
14  0 days 13:00:00                 NaT     NaN
15  0 days 21:50:00                 NaT     NaN
16  0 days 14:00:00                 NaT     NaN
17  0 days 11:24:00                 NaT     NaN
18  0 days 00:30:00                 NaT     NaN
19  1 days 09:30:00                 NaT     NaN

In [19]:
from sqlalchemy import create_engine

engine = create_engine('postgresql://postgres:TraCak45T@localhost/flask')

table_name = 'eat'
df.to_sql(table_name, engine, if_exists='append', index=False)


518